File: apps/hakka/Sources/Core/Detail/Llm/AnthropicStreamAssembler.swift
import Foundation
import HakkaCommon

/// Assembles an Anthropic Messages API event stream into its final message:
/// `content_block_delta` text deltas join into text, a `tool_use` block's
/// `input_json_delta` fragments concatenate into its input JSON (blocks are
/// indexed, so interleaved blocks stay separate), `message_start` carries
/// the model, and `message_delta` the stop reason. Unrelated events (`ping`,
/// unknown types) are skipped, never thrown on.
public func assembleAnthropicStream(_ events: [SseEvent]) -> AssembledStream {
    var text = ""
    var finishReason: String?
    var model: String?
    var blocks: [Int: ToolBlockSlot] = [:]

    for event in events {
        guard let payload = payloadObject(event.data) else { continue }

        switch payload["type"] as? String {
        case "message_start":
            if model == nil,
               let message = payload["message"] as? [String: Any],
               let reportedModel = message["model"] as? String {
                model = reportedModel
            }
        case "content_block_start":
            startToolBlock(payload, into: &blocks)
        case "content_block_delta":
            applyDelta(payload, into: &blocks, text: &text)
        case "message_delta":
            if let delta = payload["delta"] as? [String: Any],
               let stopReason = delta["stop_reason"] as? String {
                finishReason = stopReason
            }
        default:
            break
        }
    }

    let toolCalls = blocks.values
        .sorted { $0.slot < $1.slot }
        .map { AssembledToolCall(id: $0.id, name: $0.name, arguments: $0.arguments) }

    return AssembledStream(
        eventCount: events.count,
        text: text,
        toolCalls: toolCalls,
        finishReason: finishReason,
        model: model
    )
}

/// One tool-block accumulator, keyed by the stream's own content-block
/// index — fragments for the same block share it.
private struct ToolBlockSlot {
    var slot: Int
    var id: String?
    var name: String?
    var arguments: String
}

private func startToolBlock(_ payload: [String: Any], into blocks: inout [Int: ToolBlockSlot]) {
    guard let block = payload["content_block"] as? [String: Any],
          block["type"] as? String == "tool_use"
    else { return }
    let slot = (payload["index"] as? Int) ?? blocks.count
    blocks[slot] = ToolBlockSlot(
        slot: slot,
        id: block["id"] as? String,
        name: block["name"] as? String,
        arguments: ""
    )
}

private func applyDelta(_ payload: [String: Any], into blocks: inout [Int: ToolBlockSlot], text: inout String) {
    guard let delta = payload["delta"] as? [String: Any] else { return }
    if delta["type"] as? String == "text_delta", let chunk = delta["text"] as? String {
        text += chunk
    } else if delta["type"] as? String == "input_json_delta",
              let fragment = delta["partial_json"] as? String,
              let slot = payload["index"] as? Int,
              blocks[slot] != nil {
        blocks[slot]!.arguments += fragment
    }
}

private func payloadObject(_ data: String) -> [String: Any]? {
    guard let bytes = data.data(using: .utf8),
          let parsed = try? JSONSerialization.jsonObject(with: bytes),
          let object = parsed as? [String: Any]
    else { return nil }
    return object
}

File: packages/hakka-browser/src/ui/llm/anthropicStreamAssembler.ts
/**
 * Assembles an Anthropic Messages API event stream into its final message:
 * `content_block_delta` text deltas join into `text`, a `tool_use` block's
 * `input_json_delta` fragments concatenate into its input JSON (blocks are
 * indexed, so interleaved blocks stay separate), `message_start` carries the
 * model, and `message_delta` the stop reason. Unrelated events (`ping`,
 * unknown types) are skipped, never thrown on.
 */
import type { SseEvent } from 'hakka-core'

import type { AssembledStream, AssembledToolCall } from './assembledStream'

interface ToolBlockAccumulator extends AssembledToolCall {
  /** The stream's own content-block index — fragments for the same block share it. */
  slot: number
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}

/** Join `events` (already parsed SSE records) into the assembled Anthropic message. */
export function assembleAnthropicStream(events: SseEvent[]): AssembledStream {
  let text = ''
  let finishReason: string | undefined
  let model: string | undefined
  const blocks = new Map<number, ToolBlockAccumulator>()

  for (const event of events) {
    let payload: Record<string, unknown>
    try {
      const parsed: unknown = JSON.parse(event.data)
      if (!isRecord(parsed)) continue
      payload = parsed
    } catch {
      continue
    }

    switch (payload.type) {
      case 'message_start': {
        const message = isRecord(payload.message) ? payload.message : undefined
        if (model === undefined && typeof message?.model === 'string') model = message.model
        break
      }
      case 'content_block_start': {
        const block = isRecord(payload.content_block) ? payload.content_block : undefined
        if (block?.type !== 'tool_use') break
        const slot = typeof payload.index === 'number' ? payload.index : blocks.size
        blocks.set(slot, {
          slot,
          id: typeof block.id === 'string' ? block.id : undefined,
          name: typeof block.name === 'string' ? block.name : undefined,
          arguments: '',
        })
        break
      }
      case 'content_block_delta': {
        const delta = isRecord(payload.delta) ? payload.delta : undefined
        if (delta?.type === 'text_delta' && typeof delta.text === 'string') {
          text += delta.text
        } else if (delta?.type === 'input_json_delta' && typeof delta.partial_json === 'string') {
          const slot = typeof payload.index === 'number' ? payload.index : -1
          const block = blocks.get(slot)
          if (block) block.arguments += delta.partial_json
        }
        break
      }
      case 'message_delta': {
        const delta = isRecord(payload.delta) ? payload.delta : undefined
        if (typeof delta?.stop_reason === 'string') finishReason = delta.stop_reason
        break
      }
      default:
        break
    }
  }

  const toolCalls: AssembledToolCall[] = [...blocks.values()]
    .sort((a, b) => a.slot - b.slot)
    .map((block) => {
      const publicCall: AssembledToolCall = { arguments: block.arguments }
      if (block.id !== undefined) publicCall.id = block.id
      if (block.name !== undefined) publicCall.name = block.name
      return publicCall
    })

  return {
    eventCount: events.length,
    text,
    toolCalls,
    finishReason,
    model,
  }
}
